vulcan/lib/server/workflows/transform_data.cwl

```diff
@@ -0,0 +1,48 @@
+cwlVersion: v1.1
+class: Workflow
+
+inputs:
+  1_Data_Source_magma_query__queryTerms:
+    type: string
+    label: "query input"
+    doc: "The set of terms for constructing a magma query towards the data of interest.  Suggestion: Use the Query page of Timur to build the proper dataframe, and then simply copy over the 'query'-chunk presented in a green box there!"
+  1_Data_Source_magma_query__user_columns:
+    type: string
+    label: "user_columns input"
+    doc: "Optional comma-separated set of *quoted* strings for renaming columns of the returned data.  Suggestion: Use the Query page of Timur to build the proper dataframe, and then simply copy over the 'user_columns'-chunk presented there!"
+  1_Data_Source_magma_query__expand_matrices:
+    type: boolean
+    label: "expand_matrices input"
+    default: true
+    doc: "Whether to expand matrix attributes into individual columns, with one matrix data point per column.  In most cases, you'll want to leave this checked."
+
+outputs:
+  the_plot:
+    type: File
+    outputSource: transform_data/data
+
+steps:
+  get_data:
+    run: scripts/VIZ_query_df.cwl
+    label: 'Fetch Data'
+    in:
+      queryTerms: 1_Data_Source_magma_query__queryTerms
+      user_columns: 1_Data_Source_magma_query__user_columns
+      expand_matrices: 1_Data_Source_magma_query__expand_matrices
+    out: [data_frame]
+  transform_data:
+    run: ui-queries/data-transformation.cwl
+    label: 'Transform your data'
+    doc: "Manipulate your data frame as needed."
+    in:
+      data_frame: get_data/data_frame
+    # data will be a Hash with keys for "data" and "source_data".
+    #   `data` is the rendered, final value after parsing formulas.
+    #   `source_data` is the original values, including formula text.
+    out: [data]
```

Comment:
Oh man, was going to start fitting this in with the VizWorkflows, but this output structure doesn't play well with the VizUI as is.

VizUI-side: The VizUI steps (with `run: ui-queries/<bar/y/scatter>-plotly.cwl`) expect to have an input that is the json-data_frame given directly.

Here: stepUserInputs output their jsx `prop.value` at commit => as their output, so as the cwl `out: <output_name>`. Thus, if we ever want the source_data available for this stepUI, the way to accomplish that is indeed via the hash output structure currently implemented.  Thus, VizUI needs `in: data_frame: <transformation_step_name>/data**['data']**`... which I don't believe is valid CWL.

Paths forward:
1) I can build a catch into the VizUI visualization.tsx (essentially `const data_frame = (Object.keys(data['data_frame]) == ['data', 'source_data']) ? data['data_frame']['data'] : data['data_frame']`).
2) We build a secondary output method for stepUIs where perhaps when they have a `prop.multi_out=true` (false by default) the values of `prop.value` are spread into multiple outputs named by their keys? A.k.a. a way where line 42 could become `out: [data, source_data]`, and then we can pass `in: data_frame: <transformation_step_name>/data` to a VizUI step and it would be in the currently expected format.

I'm not sure if 2 is possible (can poke!) but I prefer it to 1 as the CWL is definitely the better place for this logic to go.  1 is a cumbersome hack that will likely become unmanageable in the future as we'd accumulate more and more instances of output logic from stepUI-X transferring into the jsx of stepUI-Y!

Reply:
Hm, yeah, I'm not sure how to do #2 with the current architecture, since we made an assumption that UI inputs would have a single output. I will poke around, I like your idea to include a flag to expand a passed-in hash value. Should be doable, and yeah, that would match better with how the back-end allows multiple outputs.

But, I want to focus a bit on making progress on the figures branch first, though, since I think that might take me awhile to finalize...I've currently broken it while fixing the TypeScript issues :stuck_out_tongue_closed_eyes:

Reply:
#2 might not be too hard, actually.  Technically, the server already supports multi outputs, the trick is teaching the frontend UI components to support multiple outputs.  It may be possible by making a deep cut into the onChange handlers.